File: bulk_credit_note/public/js/bulk_credit_note.js
console.log("Bulk Credit Note JS Loaded");

frappe.ui.form.on("Bulk Credit Note", {

    onload: function(frm) {

        // Auto set company from Global Defaults if empty
        if (!frm.doc.company) {

            frappe.call({
                method: "frappe.client.get_single_value",
                args: {
                    doctype: "Global Defaults",
                    field: "default_company"
                },
                callback: function(r) {
                    if (r.message) {
                        frm.set_value("company", r.message);
                    }
                }
            });

        }

    },

    refresh: function(frm) {
        console.log("Parent form loaded");
    },

    scan_sales_invoice: function(frm) {

        let invoice = frm.doc.scan_sales_invoice;

        if (!invoice) return;

        // Prevent duplicate invoices
        let exists = (frm.doc.items || []).some(row => row.sales_invoice === invoice);

        if (exists) {
            frappe.msgprint("Sales Invoice already added.");
            frm.set_value("scan_sales_invoice", "");
            return;
        }

        frappe.call({
            method: "bulk_credit_note.api.bulk_credit_note.get_sales_invoice_data",
            args: {
                invoice_no: invoice
            },
            callback: function(r) {

                if (!r.message) {
                    frappe.msgprint("Sales Invoice not found");
                    frm.set_value("scan_sales_invoice", "");
                    return;
                }

                // Add row in child table
                let row = frm.add_child("items");

                row.sales_invoice = invoice;
                row.customer = r.message.customer;
                row.posting_date = r.message.posting_date;
                row.grand_total = r.message.grand_total;
                row.outstanding_amount = r.message.outstanding_amount;
                row.update_stock = r.message.update_stock;

                frm.refresh_field("items");

                // Set parent posting date from first invoice
                if (frm.doc.items.length === 1) {
                    frm.set_value("posting_date", r.message.posting_date);
                }

                // Clear scan field for next scan
                frm.set_value("scan_sales_invoice", "");

                // Focus again for next scan
                frm.fields_dict.scan_sales_invoice.$input.focus();

            }
        });

    }

});


frappe.ui.form.on("Bulk Credit Note Item", {
    sales_invoice: function(frm, cdt, cdn) {

        let row = locals[cdt][cdn];
        if (!row.sales_invoice) return;

        frappe.call({
            method: "bulk_credit_note.api.bulk_credit_note.get_sales_invoice_data",
            args: {
                invoice_no: row.sales_invoice
            },
            callback: function(r) {

                if (!r.message) return;

                frappe.model.set_value(cdt, cdn, "customer", r.message.customer);
                frappe.model.set_value(cdt, cdn, "posting_date", r.message.posting_date);
                frappe.model.set_value(cdt, cdn, "grand_total", r.message.grand_total);
                frappe.model.set_value(cdt, cdn, "outstanding_amount", r.message.outstanding_amount);

                // Set update stock on child row
                frappe.model.set_value(cdt, cdn, "update_stock", r.message.update_stock);

                // Only set posting date on parent
                if (frm.doc.items.length === 1) {
                    frm.set_value("posting_date", r.message.posting_date);
                }
            }
        });
    }
});

File: bulk_credit_note/api/bulk_credit_note.py
import frappe
from erpnext.controllers.sales_and_purchase_return import make_return_doc


@frappe.whitelist()
def get_sales_invoice_data(invoice_no):
    """
    Fetch Sales Invoice details for child table autofill.
    """

    if not invoice_no:
        return {}

    if not frappe.db.exists("Sales Invoice", invoice_no):
        frappe.throw("Sales Invoice not found")

    si = frappe.get_doc("Sales Invoice", invoice_no)

    return {
        "customer": si.customer,
        "posting_date": si.posting_date,
        "grand_total": si.grand_total,
        "outstanding_amount": si.outstanding_amount,
        "update_stock": si.update_stock
    }


def create_credit_notes_on_submit(doc, method=None):
    """
    Trigger background job for bulk credit note creation
    """

    frappe.enqueue(
        "bulk_credit_note.api.bulk_credit_note.process_bulk_credit_notes",
        docname=doc.name,
        queue="long",
        timeout=600
    )

    frappe.msgprint("Bulk Credit Note processing started in background.")


def process_bulk_credit_notes(docname):
    """
    Background job to generate credit notes
    """

    doc = frappe.get_doc("Bulk Credit Note", docname)

    created_credit_notes = []

    for row in doc.items:

        if not row.sales_invoice:
            continue

        if not frappe.db.exists("Sales Invoice", row.sales_invoice):
            frappe.log_error(f"Sales Invoice {row.sales_invoice} not found")
            continue

        si = frappe.get_doc("Sales Invoice", row.sales_invoice)

        # ✅ Prevent duplicate credit note
        # existing = frappe.db.exists(
        #     "Sales Invoice",
        #     {
        #         "return_against": si.name,
        #         "is_return": 1,
        #         "docstatus": 1
        #     }
        # )

        existing = frappe.db.get_value(
            "Sales Invoice",
            {"return_against": si.name, "is_return": 1, "docstatus": 1},
            "name"
        )

        if existing:
            frappe.log_error(f"Credit Note already exists for {si.name}")
            continue

        try:
            # ✅ Create return doc
            credit_note_data = make_return_doc("Sales Invoice", si.name)
            credit_note = frappe.get_doc(credit_note_data)

            credit_note.is_return = 1
            credit_note.return_against = si.name

            credit_note.posting_date = frappe.utils.today()
            credit_note.posting_time = frappe.utils.nowtime()

            credit_note.company = doc.company
            credit_note.update_stock = row.update_stock

            # ✅ Warehouse handling
            if row.update_stock:

                if not si.set_warehouse:
                    frappe.log_error(
                        f"Sales Invoice {si.name} missing Source Warehouse"
                    )
                    continue

                credit_note.set_warehouse = si.set_warehouse

            # ==========================================================
            # 🔥 DISCOUNT FIX (INTEGRATED - NO SERVER SCRIPT NEEDED)
            # ==========================================================

            if si.discount_amount:

                # Step 1
                credit_note.calculate_taxes_and_totals()

                # Step 2
                credit_note.apply_discount_on = "Grand Total"
                credit_note.discount_amount = -abs(si.discount_amount)
                credit_note.additional_discount_percentage = (
                    si.additional_discount_percentage
                )

                # Step 3 CRITICAL: Recalculate after applying discount
                credit_note.calculate_taxes_and_totals()

                # Step 4 (prevent override during submit)
                credit_note.flags.ignore_pricing_rule = True
                

            # ==========================================================

            # ✅ Save + Submit
            credit_note.insert(ignore_permissions=True)
            credit_note.submit()

            created_credit_notes.append(credit_note.name)

        except Exception as e:
            # frappe.log_error(
            #     message=frappe.get_traceback(),
            #     title=f"Bulk Credit Note Error for {si.name}"
            # )

            frappe.log_error(
                message=frappe.get_traceback(),
                title=f"Bulk Credit Note Error | Invoice: {si.name} | Row: {row.idx}"
            )

    # ✅ Notify user
    if created_credit_notes:
        frappe.publish_realtime(
            "msgprint",
            {
                "message": f"Credit Notes Created: {', '.join(created_credit_notes)}"
            }
        )